libpod/image/layer_tree.go

```diff
@@ -0,0 +1,211 @@
+package image
+
+import (
+	"context"
+
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+// layerTree is an internal representation of local layers.
+type layerTree struct {
+	// nodes is the actual layer tree.
+	nodes map[string]*layerNode
+	// ociCache is cache for Image.ID -> OCI Image.
+	ociCache map[string]*ociv1.Image
+}
+
+// node returns a layerNode for the specified layer.
+func (t *layerTree) node(layer string) *layerNode {
+	node, exists := t.nodes[layer]
+	if !exists {
+		node = &layerNode{}
+		t.nodes[layer] = node
+	}
+	return node
+}
+
+// toOCI returns an OCI image for the specified image.
+func (t *layerTree) toOCI(ctx context.Context, i *Image) (*ociv1.Image, error) {
+	var err error
+	oci, exists := t.ociCache[i.ID()]
+	if !exists {
+		oci, err = i.ociv1Image(ctx)
+		t.ociCache[i.ID()] = oci
+	}
+	return oci, err
+}
+
+// layerNode is a node in a layerTree.  It's ID is the key in a layerTree.
+type layerNode struct {
+	children []*layerNode
+	images   []*Image
+	parent   *layerNode
+}
+
+// layerTree extracts a layerTree from the layers in the local storage and
+// relates them to the specified images.
+func (ir *Runtime) layerTree() (*layerTree, error) {
+	layers, err := ir.store.Layers()
+	if err != nil {
+		return nil, err
+	}
+
+	images, err := ir.GetImages()
+	if err != nil {
+		return nil, err
+	}
+
+	tree := layerTree{
+		nodes:    make(map[string]*layerNode),
+		ociCache: make(map[string]*ociv1.Image),
+	}
+
+	// First build a tree purely based on layer information.
+	for _, layer := range layers {
+		node := tree.node(layer.ID)
+		if layer.Parent == "" {
+			continue
+		}
+		parent := tree.node(layer.Parent)
+		node.parent = parent
+		parent.children = append(parent.children, node)
+	}
+
+	// Now assign the images to each (top) layer.
+	for i := range images {
+		img := images[i] // do not leak loop variable outside the scope
```

Comment:
Does this actually make a difference? There isn’t a closure that would survive the loop iteration.

Reply:
I don’t think so; that line reads the contents of the `img` variable _at that point_ and appends the pointer to `node.images`. If `img` changes later, that doesn’t affect `node.images`.

The copy/new scope is only necessary when the evaluation of a variable is _deferred_ after the iteration terminates, e.g. `runOneSecondLater(func() { consume(img) })`